mmaction/core/evaluation/average_precision.py

```diff
@@ -0,0 +1,337 @@
+import os.path as osp
+import pickle
+from collections import defaultdict
+
+import numpy as np
+
+from .bbox_overlaps import iou2d, spatio_temporal_iou3d, spatio_temporal_nms3d
+
+
+def pr_to_ap(precision_recall):
+    """Compute AP given precision-recall.
+
+    Args:
+        precision_recall (np.ndarray): precision_recall is an Nx2 array with
+            first column being precision and second column being recall.
+
+    Returns:
+        np.ndarray: The result of average precision.
+    """
+
+    recall_diff = precision_recall[1:, 1] - precision_recall[:-1, 1]
+    precision_sum = precision_recall[1:, 0] + precision_recall[:-1, 0]
+
+    return np.sum(recall_diff * precision_sum * 0.5)
+
+
+def frame_mean_ap(det_results, labels, videos, gt_tubes, threshold=0.5):
+    """Calculate frame mAP for tubes.
+
+    Args:
+        det_results (np.ndarray): Detection results for each frames.
```

Comment:
Maybe explain each element in `det_results` here. I search the keyword 'det_results', 'det_file' and find no explanation. As many lines directly use indices such as `det_results[:, 2]`, `det_result[j, 4:8]`, it can be confusing for beginners.